### ISSUE TITLE
Exception: java.util.ConcurrentModificationException at com.onesignal.OneSignalStateSynchronizer.generateJsonDiff(OneSignalStateSynchronizer.java:74)

### ISSUE THREAD

**Issue by DimaKoz** [OneSignal/OneSignal-Android-SDK#49]
I got the [crashlog](https://github.com/one-signal/OneSignal-Android-SDK/files/253650/log.txt) in OneSignal-Android-SDK 2.3.0. 
About devices:
<img width="936" alt="screen shot 2016-05-08 at 1 15 02" src="https://cloud.githubusercontent.com/assets/3092276/15094873/b3794a56-14c2-11e6-970b-d786ba3a1f93.png">

**Comment by jkasten2** [comment]
@DimaKoz Thanks for the crash report. It could be an issue with our SDK trying to generate another network request while processing another. Can you share a list of all thread stacks here or with support@onesignal.com? We should be confirm that is the issue from this information.

**Comment by OlegTarashkevich** [comment]
Hi! I got the same issue.

```
Fatal Exception: java.util.ConcurrentModificationException
 Raw
java.util.LinkedHashMap$LinkedHashIterator.nextEntry (LinkedHashMap.java:346)
java.util.LinkedHashMap$KeyIterator.next (LinkedHashMap.java:366)
com.onesignal.OneSignalStateSynchronizer.generateJsonDiff (OneSignalStateSynchronizer.java:74)
com.onesignal.OneSignalStateSynchronizer.access$200 (OneSignalStateSynchronizer.java:46)
com.onesignal.OneSignalStateSynchronizer$UserState.generateJsonDiff (OneSignalStateSynchronizer.java:263)
com.onesignal.OneSignalStateSynchronizer$UserState.access$000 (OneSignalStateSynchronizer.java:197)
com.onesignal.OneSignalStateSynchronizer.syncUserState (OneSignalStateSynchronizer.java:476)
com.onesignal.OneSignalStateSynchronizer$NetworkHandlerThread$1.run (OneSignalStateSynchronizer.java:440)
android.os.Handler.handleCallback (Handler.java:739)
android.os.HandlerThread.run (HandlerThread.java:61)
```
Screenshot:
[https://www.dropbox.com/s/8am1qmpd2i6uo2r/%D0%A1%D0%BD%D0%B8%D0%BC%D0%BE%D0%BA%20%D1%8D%D0%BA%D1%80%D0%B0%D0%BD%D0%B0%202016-06-07%20%D0%B2%2021.38.06.png?dl=0](url)

**Comment by jkasten2** [comment]
@OlegTarashkevich @DimaKoz 
Fixed this issue in the [2.4.2 Release](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/2.4.2).

Thanks.

**Comment by jkasten2** [comment]
Reff Issue #88, a few more occurrences of this exception were fixed in version 3.0.3.
